assemble/Construct_Momentum_Element_DG_template.F90

```diff
@@ -0,0 +1,2378 @@
+    logical :: boundary_element, turbine_face
+
+    ! What we will be adding to the matrix and RHS - assemble these as we
+    ! go, so that we only do the calculations we really need
+    real, dimension(NDIM(U), EFLOC(U,ele)) :: big_m_diag_addto,&
+         & rhs_addto
+    ! rhs terms for subcycling coming from advection bc terms
+    real, dimension(NDIM(U), NLOC(U,ele)) :: subcycle_rhs_addto
+    
+    real, dimension(NDIM(U), NDIM(U), EFLOC(U,ele), EFLOC(U,ele)) :: big_m_tensor_addto
+    logical, dimension(NDIM(U), NDIM(U)) :: diagonal_block_mask, off_diagonal_block_mask
+    ! Addto matrices for when subcycling is performed
+    real, dimension(NDIM(U), NDIM(U), EFLOC(U,ele), &
+         EFLOC(U,ele)) :: subcycle_m_tensor_addto
+
+    !Switch to select if we are assembling the primal or dual form
+    logical :: primal
+
+    ! In parallel, we assemble terms on elements we own, and those in
+    ! the L1 element halo
+    logical :: assemble_element
+
+    ! Absorption matrices
+    real, dimension(NDIM(U), NGI(U,ele)) :: absorption_gi
+    real, dimension(NDIM(U), NDIM(U), NGI(U,ele)) :: tensor_absorption_gi
+
+    ! Add vertical velocity relaxation to the absorption if present
+    real, intent(in) :: vvr_sf
+    real, dimension(NDIM(U),NDIM(U),NGI(U,ele)) :: vvr_abs
+    real, dimension(NDIM(U),NGI(U,ele)) :: vvr_abs_diag
+    real, dimension(NGI(U,ele)) :: depth_at_quads
+    type(scalar_field), intent(in) :: depth
+
+    ! Add implicit buoyancy to the absorption if present
+    real, intent(in) :: ib_min_grad
+    real, dimension(NDIM(U),NDIM(U),NGI(U,ele)) :: ib_abs
+    real, dimension(NDIM(U),NGI(U,ele)) :: ib_abs_diag
+    real, dimension(NLOC(U,ele),NGI(U,ele),MDIM(u)) :: dt_rho
+    real, dimension(NDIM(U),NGI(U,ele)) :: grav_at_quads
+    real, dimension(NDIM(U),NGI(U,ele)) :: grad_rho
+    real, dimension(NGI(U,ele)) :: drho_dz
+
+    ! Non-linear approximation to the PhaseVolumeFraction field
+    type(scalar_field), intent(in) :: nvfrac
+    type(element_type), pointer :: nvfrac_shape
+    ! Transformed gradient function for the non-linear PhaseVolumeFraction. 
+    real, dimension(:, :, :), allocatable :: dnvfrac_t
+    ! nvfrac at quadrature points.
+    real, dimension(NGI(U,ele)) :: nvfrac_gi, u_nl_dot_grad_nvfrac_gi
+    real, dimension(NDIM(U), NGI(U,ele)) :: grad_nvfrac_gi
+
+    ! element centre and neighbour centre
+    ! for IP parameters
+
+    real, dimension(NDIM(X)) :: ele_centre, neigh_centre, &
+         & face_centre, face_centre_2
+    real :: turbine_fluxfac
+
+    real, dimension(NGI(U,ele)) :: alpha_u_quad
+
+    ! added for partial stress form (sp911)
+    logical, intent(in) :: partial_stress
+
+    ! LES - sp911
+    real, intent(in) :: smagorinsky_coefficient
+    type(scalar_field), pointer, intent(inout) :: eddy_visc, y_plus_debug, &
+         & les_filter_width_debug
+    type(scalar_field), pointer, intent(in) :: prescribed_filter_width, distance_to_wall
+
+    dg=continuity(U)<0
```

Comment:
We can probably gain some more by admitting that momentum_dg() doesn't actually support CG, add a check at the beginning (outside the loop) and remove all the associated logic.